### ISSUE TITLE
[STORM-2191] shorten classpaths in worker and LogWriter commands

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5973]
<p>When launching the worker daemon and its wrapping LogWriter daemon, the commands can become so long that they eclipse the default Linux limit of 4096 bytes. That results in commands that are cut off in <tt>ps</tt> output, and prevents easily inspecting the system to see even what processes are running.</p>

<p>The specific scenario in which this problem can be easily triggered: <b>running Storm on Mesos</b>.</p>

<h5><a name="Detailsonwhyithappens%3A"></a>Details on why it happens:</h5>
<ol>
	<li>using the default Mesos containerizer instead of Docker containers, which causes the storm-mesos package to be unpacked into the Mesos executor sandbox.</li>
	<li>The <a href="https://github.com/apache/storm/blob/6dc6407a01d032483edebb1c1b4d8b69a304d81c/bin/storm.py#L114-L140" class="external-link" target="_blank" rel="nofollow noopener">"expand all jars on classpath"</a> functionality in the <tt>bin/storm.py</tt> script causes every one of the jars that storm bundles into its lib directory to be explicitly listed in the command.
	<ul>
		<li>e.g., say the mesos work dir is <tt>/var/run/mesos/work_dir/</tt></li>
		<li>and say that the original classpath argument in the supervisor cmd includes the following for the <tt>lib/</tt> dir in the binary storm package:
		<ul>
			<li><tt>/var/run/mesos/work_dir/slaves/2357b762-6653-4052-ab9e-f1354d78991b-S12/frameworks/20160509-084241-1086985738-5050-32231-0000/executors/STORM_TOPOLOGY_ID/runs/e6a1407e-73fd-4be4-8d00-e882117b3391/storm-mesos-0.1.7-storm0.9.6-mesos0.28.2/lib/*</tt></li>
		</ul>
		</li>
		<li>That leads to a hugely expanded classpath argument for the LogWriter and Worker daemons that get launched:
		<ul>
			<li><tt>/var/run/mesos/work_dir/slaves/2357b762-6653-4052-ab9e-f1354d78991b-S12/frameworks/20160509-084241-1086985738-5050-32231-0000/executors/STORM_TOPOLOGY_ID/runs/e6a1407e-73fd-4be4-8d00-e882117b3391/storm-mesos-0.1.7-storm0.9.6-mesos0.28.2/lib/asm-4.0.jar:/var/run/mesos/work_dir/slaves/2357b762-6653-4052-ab9e-f1354d78991b-S12/frameworks/20160509-084241-1086985738-5050-32231-0000/executors/STORM_TOPOLOGY_ID/runs/e6a1407e-73fd-4be4-8d00-e882117b3391/storm-mesos-0.1.7-storm0.9.6-mesos0.28.2/lib/carbonite-1.4.0.jar:...</tt></li>
		</ul>
		</li>
	</ul>
	</li>
</ol>

---
<details><summary><i>Originally reported by <a title="Erik Weathers" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-2191" target="_blank">shorten classpaths in worker and LogWriter commands</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Erik Weathers" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-05-08T07:42:12+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>, <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ptgoetz" class="user-hover" rel="ptgoetz">P. Taylor Goetz</a>, <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a>:  I'm working on fixing this issue, but I've noticed that in newer (as compared to 0.9.x) storm versions the <tt>get_jars_full</tt> logic has spread into the <a href="https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/daemon/supervisor/BasicContainer.java#L330-L356" class="external-link" target="_blank" rel="nofollow noopener">core Storm Java code</a> now too, and there's <a href="https://github.com/apache/storm/blob/285668742742da6316d38c1ef492c7c873b5a649/bin/storm.py#L131-L138" class="external-link" target="_blank" rel="nofollow noopener"><b>more</b> uses</a> of <tt>get_jars_full</tt> within <tt>bin/storm.py</tt>.   I do not understand the purpose for this logic and intend to submit a PR that replaces it with simply putting a wildcard in the dir that is being expanded.  Without this change the commands get obscenely long (as explained in the Description of this ticket).</p>

<p>Can you please let me know if this would be unacceptable?  I <b>really</b> don't want to have to forever fork and hack the core storm repo in order to support running the storm daemons from a deep directory (as is necessitated when using non-dockerized Mesos).</p>

<p>Thanks!!</p>

<ul class="alternate" type="square">
	<li>Erik</li>
</ul>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a>,</p>

<p>Sorry I didn't see your question before.  The issue with using * in a java classpath is that ordering of the jars is not consistent.  For Hadoop we have had all kinds of issues where a failure will only happen on some boxes and not on others because the jars on the classpath will be written in different orders and the underlying file system will return the files in different a different order.  We have not had too many issues with this in storm, but the operation was meant to be defensive.</p>

<p>If we really do want to reduce the command line we can probably play some games with jars like surefire does with a manifest only jar, but they automatically disable that in some situations so I don't know enough about why they are doing that.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>:  thanks for responding (I'm guessing that <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hmclouro" class="user-hover" rel="hmclouro">Hugo Da Cruz Louro</a> or <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a> reached out to you based on the timing, since I chatted with them about this ticket last Thursday after the Storm Meetup at Hortonworks).</p>

<p>The random ordering of jars on classpath is an issue for sure &#8211; ideally that's a reason to avoid conflicts in an application's jars, but I guess it's not always feasible.  I was assuming the operation <b>originally</b> was functional as opposed to defensive in this manner though, because the Storm code at that time supported Java 5 which didn't have the <tt>*</tt> wildcard classpath support, and so if you wanted jars on your classpath you had to list them explicitly.</p>

<p>Regarding "if we really do want to reduce the command line", for us I'd say that's an absolute requirement for us to support Storm. We need to be able to see what the different processes are and what the different topologies are for each process, and those fields are at the end of the commands.  <b>Really</b>, operationally I don't wanna accept having any commands that get cut off because of being over 4096 in general, but those are the main driving reasons.</p>

<p>Regarding your suggestion (manifest only jars?), I didn't quite follow how it would work nor how it could shorten the command lines.</p>

<p>Here's another random idea:  use the <tt>CLASSPATH</tt> environment variable instead of passing the classpath as a command line parameter, then we could have it fully filled out with these huge numbers of cmds.  I don't love it since it obscures from the <tt>ps</tt> output what the classpath is, but maybe it satisfies both of our requirements.   Also I'm not sure if it would work via the <tt>LogWriter</tt>; i.e., would <tt>LogWriter</tt> be able to pass along its <tt>CLASSPATH</tt> env var to the launched child <tt>Worker</tt> process, or would it need to reconstruct the <tt>CLASSPATH</tt> env variable like how the <tt>Supervisor</tt> would do under this proposal.</p>

<p>Would you have any time this week to talk about this a bit more in detail in some non-JIRA channel?  I'm happy to try to pursue your "manifest only jars" idea if I can learn more about what it means and how it could solve the problem.  My email address is embedded in my Apache JIRA user info.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a> Instead of changing classpath and if the main motivation here is to able to find the worker pid via greps or to make other operational parts easier, why not we pass a flag called "topology.name" to jvm args when supervisor starting the process. we should add these at the beginning of the java command so that they won't get overwrittent by class path.<br/>
java -Dtopology.name=TopologyName -Dstorm.service=Worker</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a> : I thought of that idea before too... one (minor?) problem with the idea is that we'll need to hope we don't make a mistake and we correctly "promote" all possibly needed info into the front of the cmdline.  e.g., I had forgotten we'll need the worker port field as well.  Oh, we'll need the topology <b>ID</b> too, not just the name.  Might be useful to also promote the worker-ID and supervisor-ID which are being passed as args to the <tt>Worker</tt> class's <tt>main()</tt> method.</p>

<p>In terms of usability of this feature, we'll have to change a bunch of operational scripts (e.g., monitors and deploy things) to somehow look for these cmdline parameters in a different fashion for different versions of Storm.  It's doable but a bit unfun.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a> one thing we could do is to allow users  to override that via topology.config option , it could be at cluster level or individual users can add it to their topology config along with a default.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a> I think you're suggesting possibly allowing the "classpath wildcarding" via an option.  Can you please confirm if I understood correctly?</p>

<p>One thing I was thinking about w.r.t. this idea of the wildcard being dangerous:  we are already exposed to funkiness from OS filesystem variances in the way that we resolve the classpaths today, are we not?  i.e., <a href="https://github.com/apache/storm/blob/7e183c010ee1ab342c923596d0474e4093571a43/bin/storm.py#L121" class="external-link" target="_blank" rel="nofollow noopener"><tt>get_jars_full</tt></a> and <a href="https://github.com/apache/storm/blob/7e183c010ee1ab342c923596d0474e4093571a43/storm-server/src/main/java/org/apache/storm/daemon/supervisor/BasicContainer.java#L342" class="external-link" target="_blank" rel="nofollow noopener"><tt>getFullJars</tt></a> are using OS filesystem calls to get the list of files.  So I'm not sure how it would be <b>worse</b> to use wildcards.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>I really would prefer to not have more configs.  Especially for something like this.  The * classpath is desirable, but we just need a way to be sure that there are no issues with it in the classpath.  I will see if I can put together a script so we can check.  we might be able to do it as part of our integration testing so we know that it will not be an issue with a release.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p>OK I have the script, at least a first go at it.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">ls ./apache-storm-2.0.0-SNAPSHOT/lib-worker/*.jar | xargs -n 1 jar -tf | grep <span class="code-quote">'\.class$'</span> | sort | uniq -c | grep -v <span class="code-quote">'^ *1 '</span>
</pre>
</div></div>

<p>It looks for any .class file in all of the jars with more then one instance.</p>

<p>Right now on trunk we appear to have an issue with kryo.  If we want to use the * we should fix our issues with overlaps, or decide that they are safe and we can ignore them, then make a script to verify our assumptions with each build.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha">sriharsha</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a> <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a> my suggestion of config is not to do with the classpath. I was addressing <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a> concern about passing JVM params that can be greppable,</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">one (minor?) problem with the idea is that we<span class="code-quote">'ll need to hope we don'</span>t make a mistake and we correctly <span class="code-quote">"promote"</span> all possibly needed info into the front of the cmdline. e.g., I had forgotten we<span class="code-quote">'ll need the worker port field as well. Oh, we'</span>ll need the topology ID too, not just the name. Might be useful to also promote the worker-ID and supervisor-ID which are being passed as args to the Worker <span class="code-keyword">class'</span>s main() method.
</pre>
</div></div>
<p>To make the serviceName configurable we can use a topology level config that can be set at site-wide in storm.yaml or pass through topology config itself. So that we are not hard-coding this into supervisor worker-start and also allowing users to set what is desired in here.<br/>
This will address the concerns raised in this jira without having any changes to classpath.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>: so the existing conflicting classes in <tt>lib-worker</tt> are arising between <tt>kryo-shaded-3.0.3.jar</tt> and 3 other jars:</p>
<ul>
	<li><tt>asm-5.0.3.jar</tt></li>
	<li><tt>kryo-3.0.3.jar</tt></li>
	<li><tt>reflectasm-1.10.1.jar</tt></li>
</ul>


<p>I'm not sure what the purpose of kryo-shaded <b>is</b>.  It's including stuff from all these jars, but most aren't actually shaded into different package paths than the original package paths (in <tt>kryo-3.0.3.jar</tt> and <tt>reflectasm-1.10.jar</tt>).  The only classes that are shaded into different package paths are some of the objectweb ones from <tt>asm-5.0.3.jar</tt>, but not all classes from that jar are included into kryo-shaded... i.e., these aren't included:  </p>
<ul>
	<li><tt>org/objectweb/asm/TypeReference.class</tt></li>
	<li><tt>org/objectweb/asm/signature/SignatureReader.class</tt></li>
	<li><tt>org/objectweb/asm/signature/SignatureVisitor.class</tt></li>
	<li><tt>org/objectweb/asm/signature/SignatureWriter.class</tt></li>
</ul>


<p>This conflicting-classes investigation feels like a diversion from the goal of this issue, maybe we should open a separate issue to continue discussion of these conflicts, which should be resolved anyways.</p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a> : if I'm understanding your proposal correctly, we would still be "promoting" the Worker and LogWriter java-main arguments into the front of the commands, we'd just be adding another facility that allows a topology-specific configurable string to also be "promoted" to the front of the cmds.  To be 100% clear, that topology-specific string would not suffice on its own, we still need info that is variable between worker processes and topology instances (worker port, topology ID).</p>

<p>Another wrinkle with this proposal is we should do it for all daemons really, not just Worker and LogWriter.  Nimbus, Logviewer, UI, etc. all can have these gigantic classpaths because of this classpath enumeration behavior.</p>

<p>Finally, the output of <tt>ps</tt> is so much more visually parseable when the lines aren't 4000+ characters long.  Having a command that fills an entire terminal screen is hard to work with.  I personally really dislike looking at huge hadoop and storm cmds and having to run them through colorized <tt>grep</tt> and/or <tt>tr</tt> to be able to see what is going on.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p>A few thoughts:</p>

<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>:  I dug a bit into the kryo clashes and it's because of using <tt>carbonite</tt> but also having a direct dependency on <tt>kryo</tt>.  Though carbonite's kryo dependency gets ignored because of our own kryo dependency existing, carbonite's dependency on <tt>chill-java</tt> pulls in <tt>kryo-shaded</tt>.  kryo-shaded is just kryo with some reflectasm stuff shaded in.  I imagine we can work around this problem by either:</p>
<ol>
	<li>excluding kryo-shaded from the carbonite dependency in storm's pom. - OR -</li>
	<li>excluding kryo from the carbonite dependency and adding an explicit dependency on kryo-shaded to storm's pom.</li>
</ol>


<p>Notably, the general <tt>lib/</tt> and the <tt>lib-webapp/</tt> dependencies show another conflict when I run <a href="https://issues.apache.org/jira/browse/STORM-2191?focusedCommentId=15982927&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15982927" class="external-link" rel="nofollow">Bobby's script from above</a> [1]:  <tt>javax/servlet</tt>.   That conflict arises from <tt>jetty-server</tt> doing something weird with its dependencies where it puts the <tt>javax.servlet</tt> bits under <tt>org.eclipse</tt>, but keeps the internal package path of the classes the same (doesn't relocate).  We can presumably overcome this through exclusions too.</p>

<hr />
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a> & <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sriharsha" class="user-hover" rel="sriharsha">Harsha</a>:  I still am coming back to my doubt about the "full jar" expansion being any better than the wildcard.  Both rely on running on the same host file system.  The current mechanism of doing "full jar" expansion uses both Python and Java methods to get the list of files from the filesystem. My proposed mechanism of having the JVM do the wildcard expansion would switch this to doing whatever the JVM implementation does for listing files, which would presumably be just using the same filesystem primitives.  I'm not seeing how the "full jar" expansion produces a better, more predictable ordering of the jars.</p>
<hr />
<p>Finally, I had another wacky idea:  is there any possibility of building an uber jar of storm's dependencies?  i.e., instead of all these individual 3rd party dependency jars that are being enumerated, we might be able to have a "storm-dependencies.jar" uber jar that just slaps all of the non-shaded dependencies into a single jar without relocating anything.</p>
<hr />
<p>[1] I had to slightly modify Bobby's script.  <tt>ls</tt> on my computer shows a "*" suffix for executable files, so I used <tt>find</tt> instead.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a>,</p>

<p>I thought we were sorting the jars.  But looking at the code we are not.  I am officially gobsmacked.  So my reason for not using '&#42;' was completely wrong and I am changing my vote to using the '&#42;' 100% no need for my script because as the '&#42;'  is identical to what we are doing now in terms of functionality.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a> sweet. <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>  I'm just glad that when Mr. Marz added <tt>get_jars_full</tt>, he didn't put in some sort method.  I'm hopeful this just means it was done for functionality reasons as <a href="https://issues.apache.org/jira/browse/STORM-2191?focusedCommentId=15981693&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15981693" class="external-link" rel="nofollow">I guessed above</a>.</p>

<p>I do think we should still address the conflicts we've noted above, and enable some CI test to bail out if a conflict arises, as that is probably the best way to avoid such problems arising in the future.  That can be separate, I'll file a ticket for it.  I'll send a PR soon with my changes for wildcarding.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a>,</p>

<p>Yes I think we should put them in for 1.x, but until unless we turn shading back on for master I am not concerned.  I have a pull request that will fix the issue, at least on the worker classpath.  <a href="https://github.com/apache/storm/pull/2060" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/2060</a> Although that is a side effect of removing more stuff from the worker classpath.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>:  I didn't quite follow how <a href="https://github.com/apache/storm/pull/2060" class="external-link" target="_blank" rel="nofollow noopener">PR #2060</a> fixes the conflict with kryo-shaded, can you please explain a bit?  Thanks!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a>,</p>

<p>#2060 removes clojure from the worker classpath and along with that kryo-shaded (by way of carbonite).  So now if you want to use the clojure API you need to include the dependency storm-clojure with you topology and it will ship clojure and carbonite, etc.  So it will still show up on your classpath, but we know the order of the classpath and it should not be an issue.</p>

<p>#2060 did not fix the problem for storm-server, as storm-server still has clojure as a dependency for the ui and the logviewer, but once those are moved over to java then there will be no issues with it any longer on master.  </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>: ah that makes sense, thanks for explaining!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2">revans2</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a>,</p>

<p>I merged the classpath shortening into master.  You marked this as a problem on 1.x too but the cherry-pick is not clean so if you want it there you will need a second pull request.  I'll let you close this JIRA if you are OK with just 2.0 getting this change.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>: thanks for letting me know!  I"ll prepare another PR for the 1.x branch, since we wanna upgrade to something with this change likely before 2.0 gets released.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=revans2" class="user-hover" rel="revans2">Robert Joseph Evans</a>: here's the PR for 1.x-branch:</p>
<ul>
	<li><a href="https://github.com/apache/storm/pull/2101" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/2101</a></li>
</ul>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a>, it's been merged into master and 1.x-branch.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw">erikdw</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan" class="user-hover" rel="kabhwan">Jungtaek Lim</a>: thanks for merging the change to 1.x-branch!</p>

<p>I'm not sure of the relationship between 1.x-branch and 1.0.x-branch.  I'm <b>guessing</b> that 1.x-branch is for 1.1+ versions (excluding 2.0+), and that it excludes 1.0.x versions.  Is that accurate?  I don't think this change needs to be in 1.0.x, so long as we release another version of 1.1+ at some point before 2.0+ becomes a reality (which I assume is still many months away).  i.e., I just wanna have some version with this fix before 2018. <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=erikdw" class="user-hover" rel="erikdw">Erik Weathers</a></p>

<p>That branch was targeted to 1.1.1 but I changed the target version to 1.2.0, given that this is rather improvement, not a kind of bug.<br/>
Btw, I think (and hope) we should have either 2.0.0 or 1.2.0 before 2018. If we count beta or RC, former must be happen before 2018.</p>
